File: Leetcode/1339.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
const MOD int64 = 1e9 + 7
var (
	dp map[*TreeNode]int
	sum int
    ans int64
)
func solve(root *TreeNode) int {
	val, ok := dp[root]
	if ok {
		return val
	}
	dp[root] = (*root).Val
	if (*root).Left != nil {
		dp[root] += solve((*root).Left)
	}
	if (*root).Right != nil {
		dp[root] += solve((*root).Right)
	}
	return dp[root]
}
func dfs1(root *TreeNode) {
	if root != nil {
        sum += (*root).Val
		dfs1((*root).Left)
		dfs1((*root).Right)
	}
}
func dfs2(root *TreeNode) {
	if root != nil {
		var leftSum = solve(root)
        var rightSum = sum - leftSum
        var temp = int64(leftSum) * int64(rightSum)
        if temp > ans {
            ans = temp
        }
		dfs2((*root).Left)
		dfs2((*root).Right)
	}
}
func maxProduct(root *TreeNode) int {
	dp = nil
	dp = make(map[*TreeNode]int)
	ans = math.MinInt32
	sum = 0
	dfs1(root)
	dfs2(root)
    return int(ans % MOD)
}

File: Leetcode/1339/1339.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
using ll = long long;
constexpr ll MOD = 1e9+7;

class Solution {
public:
    unordered_map<TreeNode*, int> dp; // dp[u] = the sum of the subtree with u as root
    int solve(TreeNode* u) {
        if (dp.find(u) != dp.end()) {
            return dp[u];
        }
        if (u->left && u->right) {
            dp[u] = u->val + solve(u->left) + solve(u->right);
        }
        else if (u->left && !u->right) {
            dp[u] = u->val + solve(u->left);
        }
        else if (!u->left && u->right) {
            dp[u] = u->val + solve(u->right);
        }
        else {
            dp[u] = u->val;
        }
        return dp[u];
    }
    ll ans;
    ll sum;
    void dfs(TreeNode* u) {
        if (u) {
            ll temp = solve(u);
            ll res = temp * (sum - temp);
            ans = max(ans, res);
            dfs(u->left);
            dfs(u->right);
        }
    }
    int maxProduct(TreeNode* root) {
        ans = -1;
        sum = solve(root);
        dfs(root);
        return (int)(ans % MOD);
    }
};